jupyter_server/services/kernels/handlers.py

```diff
@@ -493,116 +537,270 @@ def write_stderr(error_message):
         if channel == "iopub" and msg_type == "error":
             self._on_error(msg)
 
-        if (
-            channel == "iopub"
-            and msg_type == "status"
-            and msg["content"].get("execution_state") == "idle"
-        ):
-            # reset rate limit counter on status=idle,
-            # to avoid 'Run All' hitting limits prematurely.
-            self._iopub_window_byte_queue = []
-            self._iopub_window_msg_count = 0
-            self._iopub_window_byte_count = 0
-            self._iopub_msgs_exceeded = False
-            self._iopub_data_exceeded = False
-
-        if channel == "iopub" and msg_type not in {"status", "comm_open", "execute_input"}:
-
-            # Remove the counts queued for removal.
-            now = IOLoop.current().time()
-            while len(self._iopub_window_byte_queue) > 0:
-                queued = self._iopub_window_byte_queue[0]
-                if now >= queued[0]:
-                    self._iopub_window_byte_count -= queued[1]
-                    self._iopub_window_msg_count -= 1
-                    del self._iopub_window_byte_queue[0]
+        if self.limit_rate:
+            if (
```

Comment:
Some of this logic can be shared between the two protocols

Reply:
Yes, but I was thinking that having a separate implementation for the new protocol, and copying the code there would be enough for this PR. That way, the legacy protocol is kept unchanged, and the new protocol parses messages "progressively" as/if they are required, so it's a bit more complicated.
Also, if we were to rewrite the rate limitation logic, I would rather do it with no message parsing at all, since it's just about counting messages or bytes, there is no need to look at the message content. Again, I think it would be better to have a separate PR for that, and in that case we would share the logic between the two protocols.